### PR TITLE
Phase space layout interface

### PR DESCRIPTION
**Pull request by szabo137**
Adds the phase space layout (PSL) interface

# TODOs
- [x] consider merging after `QEDprocesses=0.3` is released

Edit: 
The remaining todos are moved to #140

### CODE REVIEW

**Comment by SimeonEhrig** [comment]
@szabo137 Sorry for the confusion. VSC likes to push on the wrong remote :roll_eyes:

**Comment by szabo137** [comment]
The fix for the unit tests are given here: https://github.com/QEDjl-project/QEDcore.jl/pull/84

File: src/interfaces/phase_space_layout.jl
Comment (AntonReinhard) [inline src/interfaces/phase_space_layout.jl:53]:
```suggestion
Return the dimensionality of the phase space, i.e., the number of coordinates, for a given
```

File: src/interfaces/phase_space_layout.jl
Comment (AntonReinhard) [inline src/interfaces/phase_space_layout.jl:65]:
```suggestion
- `proc`: The scattering process definition, a subtype of [`AbstractProcessDefinition`](@ref).
- `model`: The physics model, a subtype of [`AbstractModelDefinition`](@ref).
- `layout`: A specific phase space layout, either [`AbstractInPhaseSpaceLayout`](@ref) or
    [`AbstractOutPhaseSpaceLayout`](@ref).

```

File: src/interfaces/phase_space_layout.jl
Comment (AntonReinhard) [inline src/interfaces/phase_space_layout.jl:68]: I think we should add a note that this function should return a compile-time constant

```suggestion
## Returns
- The integer representing the number of independent phase space coordinates.

!!! note
    This function should return a compile-time constant, i.e., a number that
    can be inferred from the type information of the function call alone. If this
    is not the case, `build_momenta` and other derived functions may become
    very slow.
"""
```

File: src/interfaces/phase_space_layout.jl
Comment (AntonReinhard) [inline src/interfaces/phase_space_layout.jl:74]:
```suggestion
This function needs to be implemented for the [`AbstractOutPhaseSpaceLayout`](@ref) interface.
```

File: src/interfaces/phase_space_layout.jl
Comment (AntonReinhard) [inline src/interfaces/phase_space_layout.jl:83]:
```suggestion
- `out_psl`: The outgoing phase space layout, a subtype of [`AbstractOutPhaseSpaceLayout`](@ref).
```

File: src/interfaces/phase_space_layout.jl
Comment (AntonReinhard) [inline src/interfaces/phase_space_layout.jl:86]:
```suggestion
- The associated incoming phase space layout, a subtype of [`AbstractInPhaseSpaceLayout`](@ref).
```

File: src/interfaces/phase_space_layout.jl
Comment (AntonReinhard) [inline src/interfaces/phase_space_layout.jl:107]:
```suggestion
    - `proc`: The scattering process definition, subtype of [`AbstractProcessDefinition`](@ref).
    - `model`: The physics model, subtype of [`AbstractModelDefinition`](@ref).
    - `in_psl`: The incoming phase space layout, subtype of [`AbstractInPhaseSpaceLayout`](@ref),
```

File: src/interfaces/phase_space_layout.jl
Comment (AntonReinhard) [inline src/interfaces/phase_space_layout.jl:128]:
```suggestion
    - `proc`: The scattering process definition, subtype of [`AbstractProcessDefinition`](@ref).
    - `model`: The physics model, subtype of [`AbstractModelDefinition`](@ref).
    - `in_moms`: The incoming four-momenta, which is used to compute the momenta of the
        outgoing particles.
    - `out_psl`: The outgoing phase space layout, subtype of [`AbstractOutPhaseSpaceLayout`](@ref),
        that maps the coordinates to momenta.
```

File: src/QEDbase.jl
Comment (AntonReinhard) [inline src/QEDbase.jl:69]:
```suggestion
# Abstract phase space layout
```

File: test/interfaces/phase_space_layout.jl
Comment (AntonReinhard) [inline test/interfaces/phase_space_layout.jl:34]:
```suggestion
        @testset "in-phase-space layout" begin
```

**Review by AntonReinhard — COMMENTED** [review]
The structure and design looks good to me, as we have already talked about. I mostly have a bunch of typos and some documentation suggestions.

**Comment by AntonReinhard** [comment]
> The fix for the unit tests are given here: [QEDjl-project/QEDcore.jl#84](https://github.com/QEDjl-project/QEDcore.jl/pull/84)

I think you'll need to link the PRs via the commit messages so the tests pass and we can merge.

**Comment by szabo137** [comment]
@SimeonEhrig Sorry for my dumb question, again. As you can see in the test pipelines, this one's unit and integration tests work with a specific branch of QEDcore. Since this is not the dev branch, the verification job fails, which is intended, if I understand correctly. But now my dumb question: how should we proceed from here? I mean, this one can not be merged due to the verification failure, but https://github.com/QEDjl-project/QEDcore.jl/pull/84 can't be merged either, because it needs this PR to be merged.

**Comment by AntonReinhard** [comment]
I believe it can be merged only if you use the `CI_INTG_PKG_URL_QEDcore` comment, not with the `UNIT` one. The unit tests must succeed without a different branch, because otherwise they would fail on dev after a merge.

**Comment by szabo137** [comment]
This is understood, so that means, I need to hide the `PhaseSpaceLayout` interface behind the `QEDbase` namespace to make its unit tests pass with `QEDcore#dev` and then proceed from there?

**Comment by AntonReinhard** [comment]
> This I understood, so that means, I need to hide the `PhaseSpaceLayout` interface behind the `QEDbase` namespace to make its unit tests pass with `QEDcore#dev` and then proceed from there?

Yes, I think then it should work.

**Review by AntonReinhard — APPROVED** [review]

**Pull request merged**
